scenario_player/utils/reclaim.py

```diff
@@ -0,0 +1,520 @@
+        eth_rpc_endpoint,
+    )
+
+
+def reclaim_erc20(
+    reclamation_candidates: List[ReclamationCandidate],
+    token_address: TokenAddress,
+    contract_manager: ContractManager,
+    account: Account,
+    eth_rpc_endpoint: URI,
+):
+    web3 = Web3(HTTPProvider(eth_rpc_endpoint))
+
+    reclaim_amount = 0
+
+    log.info("Checking chain for claimable tokens", token_address=token_address)
+    for node in reclamation_candidates:
+        client = JSONRPCClient(web3, node.privkey)
+        confirmed_block_hash = client.get_confirmed_blockhash()
+        token = CustomToken(client, token_address, contract_manager, confirmed_block_hash)
+
+        balance = token.balance_of(to_canonical_address(node.address))
+        log.debug(
+            "balance",
+            token=to_checksum_address(token_address),
+            balance=balance,
+            address=node.address,
+        )
+        if balance > 0:
+            drain_amount = balance
+            log.info(
+                "Reclaiming tokens",
+                from_address=node.address,
+                amount=drain_amount.__format__(",d"),
+            )
+            reclaim_amount += drain_amount
+            assert account.address
+            try:
+                token.transfer(account.address, TokenAmount(drain_amount))
+            except InsufficientEth:
+                log.warning(
+                    "Not sufficient eth in node wallet to reclaim",
+                    address=node.address,
+                    token_address=token_address,
+                )
+                continue
+
+    if reclaim_amount:
+        log.info(
+            "Reclaimed",
+            reclaim_amount=reclaim_amount.__format__(",d"),
+            token_address=token.address,
+        )
+
+
+def reclaim_eth(
+    reclamation_candidates: List[ReclamationCandidate], account: Account, eth_rpc_endpoint: URI
+):
+    web3 = Web3(HTTPProvider(eth_rpc_endpoint))
+
+    txs = []
+    reclaim_amount = 0
+    gas_price = web3.eth.gasPrice
+    reclaim_tx_cost = gas_price * VALUE_TX_GAS_COST
+
+    log.info("Checking chain for claimable ETH")
+    for node in reclamation_candidates:
+        balance = web3.eth.getBalance(node.address)
+        if balance > reclaim_tx_cost:
+            drain_amount = balance - reclaim_tx_cost
+            log.info("Reclaiming", from_address=node.address, amount=drain_amount.__format__(",d"))
+            reclaim_amount += drain_amount
+            client = JSONRPCClient(web3, node.privkey)
+            assert account.address
+            txs.append(
+                client.transact(
+                    EthTransfer(
+                        to_address=account.address, value=drain_amount, gas_price=gas_price
+                    )
+                )
+            )
+        (node.node_dir / "reclaimed").touch()
+
+    wait_for_txs(web3, txs, 1000)
+    log.info("Reclaimed", reclaim_amount=reclaim_amount.__format__(",d"))
+
+
+def _get_all_token_network_events(
```

Comment:
I changed the previous implementation which has one fetch request per user to a single request per token network. This has a few important benefits:

- We know that when multiple nodes interact with the same Geth/Parity node using the `eth_getLogs`interface the performance degrades very fast and leads to timeout errors. If we did use the previous implementation to fetch the logs from the nodes ,we could inadvertently cause other tests to crash because the underlying Ethereum node would be under stress, and the *others* requests could fail too.
- Additionally this return the event for close channels too, which is important, since calling withdraw on a closed channel is an error.